### PR TITLE
Add incomplete events

### PR DESCRIPTION
**Pull request by benjben**
This PR adds incomplete events to Enrich flow.

Before, enriching an event [was resulting](https://github.com/snowplow/enrich/blob/4.0.0/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L69) in either a `BadRow` or an `EnrichedEvent`. What's new is that now a `BadRow` [can have](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L69) an `EnrichedEvent` associated with it, which is the incomplete event.

All the changes are related to the fact that now we don't short circuit the processing as soon as there is an error but we keep going with what is valid.

As a result we're not using `ValidatedNel` and `EitherT` any more.

To make things easier to follow, I reorganized the flow in 3 steps:
- [Mapping and validating the input](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L71).
- [Running the enrichments](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L73). This step now takes care of validating the enrichments contexts.
- [Validating the atomic fields lengths](https://github.com/snowplow/enrich/blob/incomplete_events/modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala#L89).

### CODE REVIEW

File: modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/package.scala
Comment (istreeter) [inline modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/package.scala:28]: So this can be either just a BadRow or a just an EnrichedEvent or both.  Have you ever come across [cats Ior](https://typelevel.org/cats/datatypes/ior.html)?

Comment (benjben) [reply modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/package.scala:28]: I didn't know about `Ior`. It seems like a good fit indeed. But what I don't like is that in case of an incomplete event, it is at the exact same place as a valid enriched event. I kind of like that the incomplete event is in the `Left`, coupled with the `BadRow`. So I'd rather leave it like that, except if you strongly think that we should swich to `Ior`.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (istreeter) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:171]: Would be neater and a tiny bit more efficient to do it like:

```scala
val maybeBadRow = errors.toNel.map { nel =>
  val enrichmentFailures = nel.map {
    case (schemaKey, clientError) => ???
  }
  EnrichmentManager.buildEnrichmentFailuresBadRow(.....)
}
```

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (istreeter) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:73]: Consider:

```scala
maybeBadRow = (invalidContexts ::: invalidUnstruct).toNel.map { nel =>
  buildSchemaViolationsBadRow(nel, ???, ???, ???, ???)
}
```

**Review by istreeter — COMMENTED** [review]
No real strong objections from me on this.  I do think it has made the code harder to read compared with how it was before, but maybe that's inevitable when adding a complex feature.

The point about `Ior` is that it has a Monad on the right hand side.  So it might allow arranging the code like this:

```scala
for {
  a <- doSomethingThatMightYieldBadRows(collectorPayload)
  b <- doSomethingElseThatMightYieldBadRows(a)
  event <- andSomethingElse(b)
} yield event
```

...so you accumulate failures on the left hand side, while continuing to process the event on the right hand side.

There's even a `IorT` which is similar to an `EitherT`.

I understand this work was done to unblock the next phases of this project, so I don't think we should stick around for too long searching for perfection.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:68]: That's exactly the plan! There is another ticket about adding configuration for the feature

**Comment by benjben** [comment]
Thanks for the review @istreeter ! I will address your feedback.

> I do think it has made the code harder to read compared with how it was before

Yeah I feel the same :smiling_face_with_tear: But the logic has become more complicated, so I'm afraid it's an inevitable outcome. Now after each step we need to check if there are bad rows and the value of `emitIncomplete`.

> The point about Ior is that it has a Monad on the right hand side

I'm not sure that we will benefit from it, as we need to check for bad rows after each step, and not only at the end. But I'll try `Ior` to see if it can simplify the workflow.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:252]: Way neater indeed!

**Comment by benjben** [comment]
Hi @istreeter , I addressed your feedback, please have a look. I agree that having `Ior[BadRow, EnrichedEvent]` as the return type of `enrichEvent` looks nicer. Then I tried to use `Ior` everywhere in the flow but I didn't like it. One problem is that to get the errors you need to pattern match and to have `case Ior.Left` whereas we know already that we will *never* be in this case.

> for {
>  a <- doSomethingThatMightYieldBadRows(collectorPayload)
>  b <- doSomethingElseThatMightYieldBadRows(a)
>  event <- andSomethingElse(b)
> } yield event
> ...so you accumulate failures on the left hand side, while continuing to process the event on the right hand side.

I might be missing something because I don't see how this can work. Before executing `doSomethingElseThatMightYieldBadRows` we need to check the errors that are in `a` (not everyone will have incomplete events activated), how would that work? I want to be able to accumulate the errors *and* to continue the processing (so no short circuiting with `IorT`).

An alternative that I envisaged was to use `IorT` and to have a `Ref[F, List[BadRow]]` . We would pass it to each function, along with `emitIncomplete`. Then inside each function, when there is an error, if `emitIncomplete` is set to `false`, we return a `Left` which short circuis the rest of the processing. If set to `true`, we add the bad row to the `Ref` and we return `Right`, which will continue the processing. And at the end we look at what is inside the `Ref` to decide what to return. Do you think that it would be nicer? We can jump on a call if it's unclear.

**Comment by benjben** [comment]
Hey @istreeter 

> Imagine for a second that everyone has incomplete events activated. Would that remove the need to check errors after each step, and therefore remove the part that makes it messy?

Indeed that would remove the need for that and would make the flow look nicer.

> Under what circumstances would someone have this feature disabled?

This feature will cost money as it will require an additional stream and additional apps to run, so I can imagine customers not wanting it.

@colmsnowplow should we consider that once this feature exists we will activate it for all our customers ?

> And in those cases, could we continue to process the event, but simply not emit it in the final step?

For sure we can but that would waste resources. That's not very satisfying to force features only to make the code look nicer.

**Comment by colmsnowplow** [comment]
I think I can offer a point of view here.

First just to double check I'm reading it right:

> Before executing doSomethingElseThatMightYieldBadRows we need to check the errors that are in a (not everyone will have incomplete events activated), how would that work?

I'm not 100% clear why we need to check the errors, but it sounds like it's so that we can decide not to continue processing the event if we hit an error. Just stating this in case it's not correct. :) 

> Under what circumstances would someone have this feature disabled? 

This could happen! An example - one of the customers in the research uses the JS enrichment for custom bot filtering. They specifically don't want that data to go to the db. First version of this they will most likely want to just disable it. Some customers might simply not want the additional cost of an extra stream and an extra loader.

> And in those cases, could we continue to process the event, but simply not emit it in the final step?

This would be acceptable. But perhaps we can find something cleaner. (Not sure as haven't yet carved out the time to read this PR properly)

**Comment by colmsnowplow** [comment]
> @colmsnowplow should we consider that once this feature exists we will activate it for all our customers ?

I think we were editing at the same time, think I answered that at least at first I don't think we can. @stanch, FYI, think you'll agree. :)

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (benjben) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:128]: Hey @istreeter , please note that I introduced some `recoverWith`s, as we don't want any `Left`, as it would short circuit the processing, even though incomplete events are enabled

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (istreeter) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:290]: I think you don't need this, although I know copied it from me!  EitherT [already has a method](https://github.com/typelevel/cats/blob/v2.10.0/core/src/main/scala/cats/data/EitherT.scala#L707) called `toIor`.  I missed it earlier because I was expecting it to be called `toIorT`.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (istreeter) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:228]: I think this is wrong -- and it's my fault because my version was also wrong.

Currently, if `sdjs` contains some good and some bad, then I think this returns a `Left` not a `Both`.  I guess you want it to return a `Both`?

Furthermore... if sdjs contains only bad, then I think it would be more logical if this function can return a `Both(failures, Nil)`.  And if you can get that working, then I think you can remove the `recoverWith`s further up the page.

**Review by istreeter — COMMENTED** [review]
There are lots of places where you've changed to emitting a SchemaViolation instead of an EnrichmentFailure.  This definitely needs checking with others!

It is a design decision that affects how users interact with the bad data. It deserves a discussion beyond just me and you in a code review.

**Review by istreeter — APPROVED** [review]

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (istreeter) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:228]: Sharing this in case you like it:

```scala
    sdjs.map { sdj =>
      check(client, sdj, registryLookup)
        .map(_ => List(sdj))
        .leftMap(NonEmptyList.one)
        .toIor
        .recoverWith { case errors => IorT.fromIor[F](Ior.Both(errors, Nil)) }
    }.foldA
```

File: modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala
Comment (benjben) [reply modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala:193]: This gets removed [here](https://github.com/snowplow/enrich/pull/874/files#diff-4381e8d9cbe4dd5147c2491673ea2686da148ae9782c0d228e59ea766981e8e4R195).

Now that this PR is approved, and that the other one about the configuration is also approved, I think that I'll just merge it in this one. This PR will be the main PR for incomplete events and all the new stuff will get merged into it as we go. And when we're fully done with the feature we'll merge it to `develop`

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/utils/IgluUtils.scala:228]: Yes I like it, it's fewer letters :laughing:

File: modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala
Comment (colmsnowplow) [reply modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala:193]: So if I understand correctly, this branch is the working branch, other work gets based off this one, until we're ready to merge to develop.

Sounds good to me!

Comment (benjben) [reply modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala:193]: Yes exactly :+1:

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:112]: Turns out that it was `.head` ! Testing FTW

File: modules/common/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/EnrichmentManagerSpec.scala
Comment (benjben) [inline modules/common/src/test/scala/com.snowplowanalytics.snowplow.enrich.common/enrichments/EnrichmentManagerSpec.scala:190]: Interestingly the order is not the same when combining several `ValidatedNel[A, B]` and several `Ior[List[A], B]`

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (spenes) [inline modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:87]: I think this comment should be removed because next two lines should be performed in all cases not just when `emitIncomplete` is enabled because `contexts` and `unstruct_event` might need to be updated if there is self describing json with superseded schema.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (colmsnowplow) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: I don't remember the discussions we had about it early on. 

I think my position on these design aspects was that we should make a best effort for this kind of thing but the implementation difficulty also has a role to play.

If it's not difficult to have it so that the tstamp is different for each, then I would say do it. If it is, then I would say let @stanch worry about whether or not that requirement becomes important enough to spend time on later.

@benjben I would defer to your call on it either way

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/AtomicFields.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/AtomicFields.scala:44]: Same, not sure what happened. Let me put it back as it was

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala:89]: Yes it's deliberate, but thanks for asking !

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: I can't find where but it was agreed that we would use the same timestamp everywhere to simplify the implementation and because we're not sure that having different timestamps is important.

Comment (istreeter) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: Do we have any examples yet of how this timestamp will be used at all?  e.g. any recommended sql query that might use this timestamp for recoveries?

Sorry to keep harping on about this timestamp.  If find it interesting from the point of view of schema design -- and how this timestamp complements the existing `etl_tstamp`.

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala
Comment (istreeter) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala:136]: Just an idea... I would have more confidence in the maintainability of this code if we changed it to be more explicit.

```scala
val message = err match {
  case ParseError.InvalidSchemaVer => "Invalid schema version"
  case ParseError.InvalidIgluUri => "Invalid Iglu URI"
  case ParseError.InvalidData => "Invalid data payload"
  case ParseError.InvalidSchema => "Invalid schema"
  case ParseError.InvalidMetaschema => "Invalid metashema"
}
```

Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/Failure.scala:136]: Much better indeed! Thank you!

File: modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala
Comment (colmsnowplow) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: [This is the only note I have recorded about it](https://www.notion.so/keep-in-the-snow/Failure-Entity-Schema-Design-6361207b9801468ca8ac2bbf448c81f0?pvs=4#f8d99bec76b44e3383a009409b11c2ff).

I do vaguely remember the convo we had before about it - if the use case of ordering failures isn't doable then I personally don't feel massively strongly about keeping it or dropping it. 

My only note on the topic is that we should do our best to avoid future breaking changes to this scheme where possible - but I don't think this decision could cause that.

^^ That's my 2c on it anyway! @benjben you'll be much more better with this more recently so I would for sure weigh your input more heavily than mine at this point :)

Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: I'm sure that it was agreed on that it was fine to use the same timestamp everywhere, at least for now, but I can't find where on Slack, Github and Notion.

I'd rather avoid to make last minute changes like that, but let me check if it's quick enough to do.

Comment (colmsnowplow) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: Sorry if my last message was unclear - I meant to communicate that I think that was agreed, but my memory of the details of the conversation is hazy, so just offering thoughts on the topic in general.

Apologies if how I phrased it made it sound like I was suggesting we make a change now - that wasn't the intention!

Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala:123]: No worries at all, I hadn't misinterpreted what you said. But I was still curious about how easy/hard it is to use different timestamps, as we should aim at what the best is. It seems quite straightforward to do, I'm giving it a go.

Comment (benjben) [reply modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala]: I implemented the change so that each failure gets its own timestamp at the time of creation, could you have a look @istreeter please ?

With the latest implementation it was quite straightforward, which was probably not the case when we decided that it was fine to use the same.

**Comment by istreeter** [comment]
I had already approved it, but it looks even better now ✅

**Pull request merged**
